src/ngx_http_lua_shdict.c

```diff
@@ -1238,11 +1242,38 @@ ngx_http_lua_shdict_incr(lua_State *L)
     dd("shdict lookup returned %d", (int) rc);
 
     if (rc == NGX_DECLINED || rc == NGX_DONE) {
-        ngx_shmtx_unlock(&ctx->shpool->mutex);
 
-        lua_pushnil(L);
-        lua_pushliteral(L, "not found");
-        return 2;
+        if (n == 3) {
+            ngx_shmtx_unlock(&ctx->shpool->mutex);
+
+            lua_pushnil(L);
+            lua_pushliteral(L, "not found");
+            return 2;
+        }
+
+        /* add value */
+        num = value + luaL_checknumber(L, 4);
```

Comment:
The `luaL_checknumber` function may throw an exception, leaving the shm mutex locked.